### ISSUE TITLE
Cancelar automaticamente as inscrições em disciplinas de uma matrícula se ela for trancada

### ISSUE THREAD

**Issue by Carlos-Eduardo-Cabral-da-Cunha** [gems-uff/sapos#286]
Quando se efetua o trancamento da matrícula de um aluno, é preciso cancelar as inscrições nas disciplinas manualmente. As inscrições nas disciplinas devem ser canceladas automaticamente ao se trancar uma matrícula.

**Comment by braganholo** [comment]
É bom colocar uma mensagem pedindo confirmação antes de prosseguir. Algo como “o trancamento da matrícula implicará em cancelamento das inscrições das disciplinas do semestre. Tem certeza que deseja continuar?”

**Comment by leomurta** [comment]
@SAASVUUV vc ficou de dar um retorno sobre a situação dessa issue. Quando puder, posta aqui para sabermos se podemos homologar.

**Comment by leomurta** [comment]
@braganholo  coloquei em homologação. Pode testar.

**Comment by braganholo** [comment]
Testei a introdução de um trancamento em um aluno que estava inscrito em disciplinas mas deu erro. 

<img width="871" height="502" alt="Image" src="https://github.com/user-attachments/assets/1c7350b8-8d72-4bda-85a3-20f64d22cc56" />

Notem que pela mensagem de erro, que aparece acima de cada disciplina, a impressão que dá é que o trancamento foi efetuado mas o SAPOS detectou que haviam inscrições em disciplinas e isso deixava o sistema inconsistente. Saí da tela e voltei, e de fato o trancamento não foi efetuado.

Esse aluno em especial tinha bolsa. O que deve ocorrer com a alocação de bolsa quando o aluno tranca? O trancamento teria que ser barrado enquanto a alocação não fosse finalizada, já que finalizar a alocação automaticamente poderia fazer essa bolsa ser "esquecida" pela coordenação.

**Comment by braganholo** [comment]
Conversamos e decidimos que o comportamento aqui deve ser impedir a realização do trancamento quando há inscrição em disciplina no período a ser trancado. No entanto, nos testes aparecem mensagens de erro conflitantes (ver imagem do meu comentário anterior). A primeira, logo acima do período de trancamento, diz que não é possível gravar o trancamento, pois existem inscrições em disciplina, e as demais, acima de cada inscrição em disciplina no período a ser trancado, diz que não é possível gravar a inscrição pois o período está trancado. No final das contas o comportamento está correto, pois o trancamento não é realizado, mas as mensagens de erro deixam dúvidas no usuário. O ideal seria que a mensagem acima das inscrições não aparecesse. É viável isso? Elas deveriam aparecer apenas se o usuário tentar criar uma inscrição em disciplina para um semestre que está trancado.

**Comment by leomurta** [comment]
@SAASVUUV acho que a pergunta da Vanessa é para você.

**Comment by braganholo** [comment]
Está ok para ir para produção.

**Comment by leomurta** [comment]
Bug no seed após a remoção de enrollment_holds.active (commit c08d1ddc, ISSUE #286)

Ao remover a coluna active, o seed db/seeds/02.reports_notifications.rb foi ajustado nas queries "CHECK INSCRIÇÃO - Fim da validade de trancamentos" e "CHECK INSCRIÇÃO - Alunos que não se inscreveram em disciplinas" apenas apagando AND eh.active = 1, mantendo eh.year = :ano AND eh.semester = :semestre.

O problema: no modelo antigo havia uma linha por semestre trancado, então esse filtro equivalia a "trancado no semestre X". No modelo novo, um único registro cobre um intervalo (year, semester, number_of_semesters). Casar só year/semester passa a testar apenas o semestre inicial do trancamento e ignora trancamentos multi-semestre.

Exemplo: trancamento {year: 2025, semester: 2, number_of_semesters: 4} (cobre 2025.2, 2026.1, 2026.2, 2027.1). Perguntando "trancado em 2026.1?", o filtro atual não casa (a linha é 2025.2) e o aluno é tratado como não-trancado.

Sugestão de correção — testar se o semestre-alvo cai no intervalo, usando 2*year + semester como índice contíguo de semestre:

(2*eh.year + eh.semester) <= (2*:ano + :semestre)
AND (2*:ano + :semestre) < (2*eh.year + eh.semester + eh.number_of_semesters)
Observação: instâncias existentes já foram corrigidas

**Comment by leomurta** [comment]
Complementando: ao remover a coluna `active` (commit `c08d1ddc`), as queries de trancamento em `db/seeds/02.reports_notifications.rb` foram ajustadas apenas **apagando** `AND eh.active = 1`, mantendo `eh.year = :ano AND eh.semester = :semestre`.

**Correção sugerida:** testar se o semestre-alvo cai no intervalo, usando `2*year + semester` como índice contíguo de semestre:

```sql
(2*eh.year + eh.semester) <= (2*:ano + :sem)
AND (2*:ano + :sem) < (2*eh.year + eh.semester + eh.number_of_semesters)
```

#### 1) `CHECK INSCRIÇÃO - Fim da validade de trancamentos`

```sql
SELECT s.name,
    s.email,
    e.enrollment_number, eh.year, eh.semester
FROM students s, enrollments e, enrollment_holds eh
WHERE s.id = e.student_id
AND e.id = eh.enrollment_id
AND (2*eh.year + eh.semester) <= (2*:ano_semestre_anterior + :numero_semestre_anterior)
AND (2*:ano_semestre_anterior + :numero_semestre_anterior) < (2*eh.year + eh.semester + eh.number_of_semesters)
```

#### 2) `CHECK INSCRIÇÃO - Alunos que não se inscreveram em disciplinas`

Muda apenas o subselect "não está trancado":

```sql
/* não está trancado */
AND e.id NOT IN (SELECT eh.enrollment_id
                 FROM enrollment_holds eh
                 WHERE (2*eh.year + eh.semester) <= (2*:ano_semestre_atual + :numero_semestre_atual)
                 AND (2*:ano_semestre_atual + :numero_semestre_atual) < (2*eh.year + eh.semester + eh.number_of_semesters))
```

**Comment by SAASVUUV** [comment]
@leomurta, realmente, as querys como estão hoje não têm o comportamento esperado, visto que não considera que um trancamento pode durar mais de um semestre. No entanto, a primeira query 'Fim de validade de trancamentos', é referente a uma notificação ao coordenador com a lista de alunos que terminam o trancamento ao final do semestre, certo?

"Prezado Coordenador

Informamos que a validade do trancamento dos alunos abaixo relacionados termina no final do mês.
Aproveitamos para informar que todos eles foram notificados de que devem se matricular para regularizar sua situação."

Então, nesse caso não deveríamos considerar somente o último semestre de trancamento?
Corrigiria a query desse modo, selecionando assim só o último semestre trancado da matrícula:

```
SELECT s.name,
          s.email,
          e.enrollment_number, eh.year, eh.semester
      FROM students s, enrollments e, enrollment_holds eh
      WHERE s.id = e.student_id
      AND e.id = eh.enrollment_id
      AND (2*eh.year + eh.semester + eh.number_of_semesters - 1) = (2*:ano_semestre_anterior + :numero_semestre_anterior)
```

**Comment by leomurta** [comment]
Concordo, @SAASVUUV  — para a "Fim da validade de trancamentos" faz mais sentido a igualdade no último semestre do trancamento, já que a notificação é sobre trancamentos que *terminam*, não sobre estar trancado no semestre. Tua versão cobre certo os casos multi-semestre e é retrocompatível com `n=1`.

Só reforçando: a segunda query ("não se inscreveram em disciplinas") continua com o teste de intervalo, porque ali o objetivo é excluir quem está trancado *no semestre atual* (independente de quando termina):

    WHERE (2*eh.year + eh.semester) <= (2*:ano_semestre_atual + :numero_semestre_atual)
    AND (2*:ano_semestre_atual + :numero_semestre_atual) < (2*eh.year + eh.semester + eh.number_of_semesters)

Vou alinhar as instâncias em produção com essas duas versões.
